processing/src/test/java/io/druid/segment/data/ThreadSafetyAssertingAggregatorFactory.java

```diff
@@ -0,0 +1,216 @@
+package io.druid.segment.data;
+
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.segment.ColumnSelectorFactory;
+import org.junit.Assert;
+
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * An AggregatorFactory that asserts thread safety.
+ * If the delegate aggregator factory is accessed in a thread unsafe manner throws AssertionError during read.
+ */
+public class ThreadSafetyAssertingAggregatorFactory extends AggregatorFactory
+{
+  private final AggregatorFactory delegate;
+
+  public ThreadSafetyAssertingAggregatorFactory(AggregatorFactory delegate)
+  {
+    this.delegate = delegate;
+  }
+
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    final Aggregator delegate1 = delegate.factorize(metricFactory);
+    final Aggregator delegate2 = delegate.factorize(metricFactory);
+    return new Aggregator()
+    {
+      @Override
+      public void aggregate()
+      {
+        delegate1.aggregate();
+        Thread.yield();
```

Comment:
If we replace yield with Thread.sleep(1), the issue is reproduced more frequently but this slows down the test considerably as it is done for every aggregate call.